### ISSUE TITLE
eventmachine gem fails to install solaris 64 bit

### ISSUE THREAD

**Issue by CirrusThink** [eventmachine/eventmachine#611]
eventmachine keeps failing when installing on Solaris in 64 bit mode

I dont know how to invoke correctly the options at installation to see if it is possible to tell it to not use g++ but SolarisStudio CC which is more appopriate. 
So far I have been able to run some test locally in Solaris 11.2 with Solaris Studio and i can install if I add the following to the extconf.rb files, just dont know how to do on the fly

CONFIG["CC"] = "cc -m64"
CONFIG["CXX"] = “CC -m64"
CONFIG["CFLAGS"] = "-m64 -xO3 -xbuiltin=%all -xinline=auto -xprefetch=auto -xdepend -KPIC"
CONFIG["LDFLAGS"] = "-m64 -L. -L/usr/sfw/lib -R/usr/sfw/lib"
CONFIG["CPPFLAGS"] = "-I/usr/sfw/include -I/usr/include"
CONFIG["CPP"] = "cc -E"
CONFIG["CCDLFLAGS"] = “-KPIC"

dolly@vdocklab01:~$ CXX=CC ; export CXX ; gem install eventmachine
Building native extensions.  This could take a while...
ERROR:  Error installing eventmachine:
        ERROR: Failed to build gem native extension.

    /export/home/dolly/ruby64/ruby-2.1.6/bin/ruby extconf.rb
checking for rb_trap_immediate in ruby.h,rubysig.h... no
checking for rb_thread_blocking_region()... yes
checking for ruby/thread.h... yes
checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
checking for inotify_init() in sys/inotify.h... no
checking for __NR_inotify_init in sys/syscall.h... no
checking for writev() in sys/uio.h... yes
checking for rb_thread_fd_select()... yes
checking for rb_fdset_t in ruby/intern.h... yes
checking for rb_wait_for_single_fd()... yes
checking for rb_enable_interrupt()... no
checking for rb_time_new()... yes
checking for sys/event.h... no
checking for main() in -lnsl... yes
checking for main() in -lsocket... yes
checking for clock_gettime()... yes
checking for CLOCK_MONOTONIC_RAW in time.h... no
checking for CLOCK_MONOTONIC in time.h... yes
creating Makefile

make "DESTDIR=" clean

make "DESTDIR="
compiling binder.cpp
g++: error: unrecognized command line option ‘-KPIC’
make: *** [binder.o] Error 1

make failed, exit code 2

**Comment by sodabrew** [comment]
These settings appear to be in https://github.com/eventmachine/eventmachine/blob/master/ext/extconf.rb#L121-136

Was your Ruby binary built with gcc or Studio?

**Comment by CirrusThink** [comment]
ruby built with Studio

**Comment by CirrusThink** [comment]
I dont think Soalris 11.2 uses SUNWspro

after compiling with the options I added into the extconf I get another issue

Unable to load the EventMachine C extension; To use the pure-ruby reactor, require 'em/pure_ruby'
/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/gems/2.1.0/bundler/gems/eventmachine-2333335a9fbf/lib/eventmachine.rb:8:in `require': ld.so.1: ruby: fatal: relocation error: R_AMD64_PC32: file /export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/gems/2.1.0/bundler/gems/extensions/x86_64-solaris-2.11/2.1.0-static/eventmachine-2333335a9fbf/rubyeventmachine.so: symbol _init: value 0x7f0044750225 does not fit - /export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/gems/2.1.0/bundler/gems/extensions/x86_64-solaris-2.11/2.1.0-static/eventmachine-2333335a9fbf/rubyeventmachine.so (LoadError)
        from /export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/gems/2.1.0/bundler/gems/eventmachine-2333335a9fbf/lib/eventmachine.rb:8:in `<top (required)>'
        from ./initializer.rb:8:in `require'
        from ./initializer.rb:8:in `<main>'

**Comment by sodabrew** [comment]
Could you run an `irb` session and `puts RbConfig::CONFIG` and see what values it has for 'CC' and such?

**Comment by sodabrew** [comment]
Oh - got it. Should there be a different case to catch the Studio compiler?

**Comment by CirrusThink** [comment]
puts RbConfig::CONFIG
{"DESTDIR"=>"", "MAJOR"=>"2", "MINOR"=>"1", "TEENY"=>"0", "PATCHLEVEL"=>"336", "INSTALL"=>"/usr/gnu/bin/install -c", "EXEEXT"=>"", "prefix"=>"/export/home/dolly/ruby64/ruby-2.1.6", "ruby_install_name"=>"ruby", "RUBY_INSTALL_NAME"=>"ruby", "RUBY_SO_NAME"=>"ruby", "exec"=>"exec", "ruby_pc"=>"ruby-2.1.pc", "PACKAGE"=>"ruby", "BUILTIN_TRANSSRCS"=>" newline.c", "USE_RUBYGEMS"=>"YES", "MANTYPE"=>"man", "NROFF"=>"/usr/bin/nroff", "vendorarchhdrdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include/ruby-2.1.0/vendor_ruby/x86_64-solaris2.11", "sitearchhdrdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include/ruby-2.1.0/site_ruby/x86_64-solaris2.11", "rubyarchhdrdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include/ruby-2.1.0/x86_64-solaris2.11", "vendorhdrdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include/ruby-2.1.0/vendor_ruby", "sitehdrdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include/ruby-2.1.0/site_ruby", "rubyhdrdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include/ruby-2.1.0", "RUBY_SEARCH_PATH"=>"", "UNIVERSAL_INTS"=>"", "UNIVERSAL_ARCHNAMES"=>"", "configure_args"=>" '--prefix=/export/home/dolly/ruby64/ruby-2.1.6' 'CC=cc -m64'", "vendorarchdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/vendor_ruby/2.1.0/x86_64-solaris2.11", "vendorlibdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/vendor_ruby/2.1.0", "vendordir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/vendor_ruby", "sitearchdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/site_ruby/2.1.0/x86_64-solaris2.11", "sitelibdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/site_ruby/2.1.0", "sitedir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/site_ruby", "rubyarchdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/2.1.0/x86_64-solaris2.11", "rubylibdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/2.1.0", "ruby_version"=>"2.1.0", "sitearch"=>"x86_64-solaris2.11", "arch"=>"x86_64-solaris2.11", "sitearchincludedir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include/x86_64-solaris2.11", "archincludedir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include/x86_64-solaris2.11", "sitearchlibdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/x86_64-solaris2.11", "archlibdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/x86_64-solaris2.11", "libdirname"=>"libdir", "RUBY_EXEC_PREFIX"=>"/export/home/dolly/ruby64/ruby-2.1.6", "RUBY_LIB_VERSION"=>"", "RUBY_LIB_VERSION_STYLE"=>"3\t/* full */", "RI_BASE_NAME"=>"ri", "ridir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/ri", "rubysitearchprefix"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/x86_64-solaris2.11", "rubyarchprefix"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/x86_64-solaris2.11", "MAKEFILES"=>"Makefile GNUmakefile", "PLATFORM_DIR"=>"", "THREAD_MODEL"=>"pthread", "SYMBOL_PREFIX"=>"", "EXPORT_PREFIX"=>"", "COMMON_HEADERS"=>"", "COMMON_MACROS"=>"", "COMMON_LIBS"=>"", "MAINLIBS"=>"", "ENABLE_SHARED"=>"no", "DLDLIBS"=>" -lc", "SOLIBS"=>"", "LIBRUBYARG_SHARED"=>"-R /export/home/dolly/ruby64/ruby-2.1.6/lib -L/export/home/dolly/ruby64/ruby-2.1.6/lib ", "LIBRUBYARG_STATIC"=>"-R /export/home/dolly/ruby64/ruby-2.1.6/lib -L/export/home/dolly/ruby64/ruby-2.1.6/lib -lruby-static", "LIBRUBYARG"=>"-R /export/home/dolly/ruby64/ruby-2.1.6/lib -L/export/home/dolly/ruby64/ruby-2.1.6/lib -lruby-static", "LIBRUBY"=>"libruby-static.a", "LIBRUBY_ALIASES"=>"libruby.so", "LIBRUBY_SO"=>"libruby.so.2.1.0", "LIBRUBY_A"=>"libruby-static.a", "RUBYW_INSTALL_NAME"=>"", "rubyw_install_name"=>"", "EXTDLDFLAGS"=>"", "EXTLDFLAGS"=>"", "strict_warnflags"=>"", "warnflags"=>"", "debugflags"=>"-g", "optflags"=>"", "cxxflags"=>"  -g ", "cflags"=>"  -g ", "cppflags"=>"", "NULLCMD"=>":", "DLNOBJ"=>"dln.o", "INSTALLDOC"=>"all", "CAPITARGET"=>"nodoc", "RDOCTARGET"=>"rdoc", "DTRACE_GLOMMED_OBJ"=>"ruby-glommed.o", "DTRACE_OBJ"=>"probes.o", "DTRACE_EXT"=>"d", "EXECUTABLE_EXTS"=>"", "ARCHFILE"=>"", "LIBRUBY_RELATIVE"=>"no", "EXTOUT"=>".ext", "RUNRUBY_COMMAND"=>"$(MINIRUBY) $(srcdir)/tool/runruby.rb --extout=.ext $(RUNRUBYOPT)", "PREP"=>"miniruby", "BTESTRUBY"=>"$(MINIRUBY)", "CROSS_COMPILING"=>"no", "TEST_RUNNABLE"=>"yes", "rubylibprefix"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby", "setup"=>"Setup", "EXTSTATIC"=>"", "STRIP"=>"strip", "TRY_LINK"=>"", "LIBPATHENV"=>"LD_LIBRARY_PATH_64", "RPATHFLAG"=>" -R%1$-s", "LIBPATHFLAG"=>" -L%1$-s", "LINK_SO"=>"", "LIBEXT"=>"a", "DLEXT2"=>"", "DLEXT"=>"so", "LDSHAREDXX"=>"g++ -G", "LDSHARED"=>"cc -m64 -G", "CCDLFLAGS"=>" -KPIC", "STATIC"=>"", "ARCH_FLAG"=>"", "DLDFLAGS"=>"", "ALLOCA"=>"", "codesign"=>"", "POSTLINK"=>":", "WERRORFLAG"=>"", "CHDIR"=>"cd -P", "RMALL"=>"rm -fr", "RMDIRS"=>"rmdir --ignore-fail-on-non-empty -p", "RMDIR"=>"rmdir --ignore-fail-on-non-empty", "CP"=>"cp", "RM"=>"rm -f", "PKG_CONFIG"=>"pkg-config", "PYTHON"=>"", "DOXYGEN"=>"", "DOT"=>"", "DTRACE"=>"dtrace", "MAKEDIRS"=>"/usr/gnu/bin/mkdir -p", "MKDIR_P"=>"/usr/gnu/bin/mkdir -p", "INSTALL_DATA"=>"/usr/gnu/bin/install -c -m 644", "INSTALL_SCRIPT"=>"/usr/gnu/bin/install -c", "INSTALL_PROGRAM"=>"/usr/gnu/bin/install -c", "SET_MAKE"=>"", "LN_S"=>"ln -s", "NM"=>"nm", "DLLWRAP"=>"", "WINDRES"=>"", "OBJCOPY"=>"objcopy", "OBJDUMP"=>"objdump", "ASFLAGS"=>"", "AS"=>"as", "AR"=>"ar", "RANLIB"=>"ranlib", "try_header"=>"", "CC_VERSION"=>"cc -m64 -V", "COUTFLAG"=>"-o ", "OUTFLAG"=>"-o ", "CPPOUTFILE"=>"-o conftest.i", "GNU_LD"=>"no", "LD"=>"/usr/ccs/bin/ld", "GCC"=>"", "EGREP"=>"/usr/gnu/bin/grep -E", "GREP"=>"/usr/gnu/bin/grep", "CPP"=>"cc -m64 -E", "CXXFLAGS"=>"  -g ", "CXX"=>"g++", "OBJEXT"=>"o", "CPPFLAGS"=>"  ", "LDFLAGS"=>"-L. ", "CFLAGS"=>"  -g ", "CC"=>"cc -m64", "NACL_SDK_VARIANT"=>"", "NACL_SDK_ROOT"=>"", "NACL_TOOLCHAIN"=>"", "target_os"=>"solaris2.11", "target_vendor"=>"pc", "target_cpu"=>"x86_64", "target"=>"x86_64-pc-solaris2.11", "host_os"=>"solaris2.11", "host_vendor"=>"pc", "host_cpu"=>"x86_64", "host"=>"x86_64-pc-solaris2.11", "RUBY_VERSION_NAME"=>"ruby-2.1.0", "RUBYW_BASE_NAME"=>"rubyw", "RUBY_BASE_NAME"=>"ruby", "build_os"=>"solaris2.11", "build_vendor"=>"pc", "build_cpu"=>"x86_64", "build"=>"x86_64-pc-solaris2.11", "RUBY_RELEASE_DATE"=>"2015-04-13", "RUBY_PROGRAM_VERSION"=>"2.1.6", "target_alias"=>"", "host_alias"=>"", "build_alias"=>"", "LIBS"=>"-lpthread -lsocket -ldl -lcrypt -lm ", "ECHO_T"=>"", "ECHO_N"=>"", "ECHO_C"=>"\\\\c", "DEFS"=>"", "mandir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/man", "localedir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/locale", "libdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib", "psdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/doc/ruby", "pdfdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/doc/ruby", "dvidir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/doc/ruby", "htmldir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/doc/ruby", "infodir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/info", "docdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share/doc/ruby", "oldincludedir"=>"/usr/include", "includedir"=>"/export/home/dolly/ruby64/ruby-2.1.6/include", "localstatedir"=>"/export/home/dolly/ruby64/ruby-2.1.6/var", "sharedstatedir"=>"/export/home/dolly/ruby64/ruby-2.1.6/com", "sysconfdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/etc", "datadir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share", "datarootdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/share", "libexecdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/libexec", "sbindir"=>"/export/home/dolly/ruby64/ruby-2.1.6/sbin", "bindir"=>"/export/home/dolly/ruby64/ruby-2.1.6/bin", "exec_prefix"=>"/export/home/dolly/ruby64/ruby-2.1.6", "PACKAGE_URL"=>"", "PACKAGE_BUGREPORT"=>"", "PACKAGE_STRING"=>"", "PACKAGE_VERSION"=>"", "PACKAGE_TARNAME"=>"", "PACKAGE_NAME"=>"", "PATH_SEPARATOR"=>":", "SHELL"=>"/bin/sh", "archdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/2.1.0/x86_64-solaris2.11", "topdir"=>"/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/2.1.0/x86_64-solaris2.11"}
=> nil

**Comment by CirrusThink** [comment]
I think if its set to CC = cc and CXX CC then its Studio

**Comment by CirrusThink** [comment]
the KPIC I believe is producing errors though with 64 bit

**Comment by CirrusThink** [comment]
I see CXX"=>"g++" also that would have to be CC

**Comment by CirrusThink** [comment]
Is there something I should be doing ?

**Comment by sodabrew** [comment]
This seems weird in your output. You have CPP set correctly for running the Studio pre-processor (cc -E flag) but CXX is set to g++, which isn't the Studio compiler.
```
"CPP"=>"cc -m64 -E",
"CXX"=>"g++",
"CC"=>"cc -m64", 
```

Could you post the output of `cc -flags` ?

**Comment by CirrusThink** [comment]
yeah it seems to be mixed up

______________________________________________________________________________
Items within [ ] are optional. Items within < > are variable parameters.
Bar | indicates choice of literal values.
______________________________________________________________________________
-#                            Verbose mode
-###                          Show compiler commands built by driver, no compilation
-A<name[(tokens)]>            Preprocessor predicate assertion
-B[static|dynamic]            Specify dynamic or static binding
-C                            Prevent preprocessor from removing comments
-D<name[=token]>              Associate name with token as if by #define
-E                            Compile source through preprocessor only, output to stdout
-G                            Build a dynamic shared library
-H                            Print path name of each file included during compilation
-I<dir>                       Add <dir> to preprocessor #include file search path
-KPIC                         Compile position independent code with 32-bit addresses
-Kpic                         Compile position independent code
-L<dir>                       Pass to linker to add <dir> to the library search path
-M<file>                      Pass <file> mapfile to linker
-O                            Use default optimization level (-xO3)
-O<n>                         Same as -xO<n>
-P                            Compile source through preprocessor only, output to .i  file
-PIC                          Same as -KPIC
-Q{y|n}                       Emit/don't emit identification info to output file
-Qoption <prog> <o>[,<o>...]  Pass options list <o> to compilation phase <prog>
-R<dir[:dir]>                 Build runtime search path list into executable
-S                            Compile and only generate assembly code (.s)
-U<name>                      Delete initial definition of preprocessor symbol <name>
-V                            Report version number of each compilation phase
-W<c>,<arg>                   Pass <arg> to specified component <c> (a,d,h,i,l,m,p,u,0,2)
-Xc                           Compile assuming strict ANSI C conformance
-Xa                           Compile assuming ANSI C conformance, allow K & R extensions (default mode)
-Xt                           Compile assuming K & R conformance, allow ANSI C
-Xs                           Compile assuming (pre-ANSI) K & R C style code
-Xlinker <arg>                Pass <arg> to linker
-Y<c>,<dir>                   Specify <dir> for location of component <c> (a,l,m,p,0,h,i,u)
-YA,<dir>                     Change default directory searched for components
-YI,<dir>                     Change default directory searched for include files
-YP,<dir>                     Change default directory for finding libraries files
-YS,<dir>                     Change default directory for startup object files
-Zll                          Create lock_lint database files (.ll)
-ansi                         Same as -std=c89
-c                            Compile only - produce .o files, suppress linking
-dalign                       Ignored
-d{n|y}                       Dynamic [-dy] or static [-dn] option to linker
-dryrun                       Show compiler commands built by driver, no compilation
-e<arg>                       Passed to linker
-errfmt[=<a>]                 Set format of compiler diagnostic messages; <a>={error}
-errhdr[=<v>]                 Enable diagnostics from standard or user include files; <v>={%all,%none,%user,<include dir>}
-erroff[=<tags>]              Suppress warnings specified by tags; <tags>={%none, %all, <tag list>}
-errshort[=<a>]               Control amount of type details in error messages;
                              <a>={full|short|tags}
-errtags[={yes|no}]           Display messages with tags
-errwarn[=<tags>]             Treats warnings specified by tags as errors; <tags>={%none, %all, <tag list>}
-fPIC                         Same as -KPIC
-fast                         Optimize using a selection of options
-fd                           Report old-style function definitions and declarations
-features[=<a>[,<a>]]         Enable/disable C language features;
                              <a>={extinl,typeof,extensions,conststrings}
-flags                        Show this summary of compiler options
-flagsrc=<f>                  Accept command options from file <f>
-flteval[=<a>]                Evaluation of floating point expressions; <a>={any|2}
-fma=<a>                      Enable floating-point multiply-add instruction; <a>={none|fused}
-fnonstd                      Initialize floating-point hardware to non-standard preferences
-fns[={yes|no}]               Select non-standard floating point mode
-fopenmp                      Equivalent to -xopenmp=parallel
-fpic                         Same as -Kpic
-fprecision=<a>               Set FP rounding precision mode; <a>={single|double|extended}
-fround=<r>                   Select the IEEE rounding mode in effect at startup
-fsimple[=<n>]                Select floating-point optimization preferences <n>
-fsingle                      Use single-precision arithmetic (-Xt and -Xs modes only)
-fstore                       Force floating pt. values to target precision on assignment
-ftrap=<t>                    Select floating-point trapping mode in effect at startup
-g                            Compile for debugging
-g0                           Compile for debugging by dbx but allow inlining
-g1                           Compile for limited debugging by dbx
-g2                           Compile for debugging by dbx
-g3                           Compile for debugging by dbx including macros.
-gnone                        Turn off debugging information.
-h <name>                     Assign <name> to generated dynamic shared library
-i                            Passed to linker to ignore any LD_LIBRARY_PATH setting
-include <file>               Include the contents of <file> before other files
-keeptmp                      Keep temporary files created during compilation
-l<name>                      Link with library lib<name>.a or lib<name>.so
-library=sunperf              Specify linking with the performance libraries
-m32                          Set 32-bit addressing model
-m64                          Set 64-bit addressing model
-mc                           Remove duplicate strings from .comment section of output files
-mr                           Remove all strings from .comment section of output files
-mr,"string"                  Remove all strings and append "string" to .comment section
-mt[={yes|no}]                Specify options needed when compiling multi-threaded code
-native                       Optimize for the host system (-xtarget=native)
-nofstore                     Do not force floating pt. values to target precision on assignment
-o <outputfile>               Set name of output file to <outputfile>
-p                            Compile for profiling with prof
-pedantic[={yes|no}]          Strict conformance with errors/warning for non-ANSI constructs
-pic                          Same as -Kpic
-preserve_argvalues={none|simple|complete} Preserve argument values
-qp                           Compile for profiling with prof
-s                            Strip symbol table from the executable file
-shared                       Same as -G
-std={c89|c99|c11}            Specify C standard
-temp=<path>                  Use <path> as directory for temporary files
-traceback[=<a>[,<a>]]        Provide stack traceback for the abnormal termination by signals; <a>={%none|common|<signal>}
-v                            Do stricter semantic checking
-w                            Suppress compiler warning messages
-xCC                          Accept C++ style comments
-xF[=<a>[,<a>]]               Compile for later mapfile reordering and data fragmentation
-xM                           Generate makefile dependencies
-xM1                          Generate makefile dependencies, but exclude /usr/include
-xMD                          Generate makefile dependencies and compile at once
-xMMD                         Generate makefile dependencies like -xMD, but excluding standard headers
-xMF <file>                   Specify output <file> for makefile dependencies dump
-xMerge                       Merge data segment into text segment
-xO<n>                        Generate optimized code; <n>={1|2|3|4|5}
-xP                           Print prototypes for function definitions
-xaddr32[={yes|no}]           Generate binaries assuming the associated process is restricted to the lower 32bit address space
-xalias_level[=<a>]           Enable optimizations based on the specified alias_level;
                              <a>={any|basic|weak|layout|strict|std|strong}
-xanalyze=code                Same as -xprevise=yes
-xannotate[={yes|no}]         Annotate binaries for optimization and analysis
-xarch=<a>                    Specify target architecture instruction set
-xautopar                     Enable automatic loop parallelization
-xbuiltin[=<a>]               Inline system functions and intrinsics when beneficial;
                              <a>={%none|%default|%all}
-xc99[=<a>]                   Enable ISO C99 features; <a>={all,none,[no_]lib}
-xcache=<t>                   Define cache properties for use by optimizer
-xchar[=<a>]                  Treat type char as signed (s) or unsigned (u);
                              <a>={s|signed|u|unsigned}
-xchar_byte_order=<o>         Specify multi-char byte order <o> (default, high, low)
-xcheck[=<a>[,<a>]]           Generate runtime checks for error condition;
                              <a>={stkovf,stkovf:diagnose,stkovf:detect,init_local}
-xchip=<a>                    Specify the target processor for use by the optimizer
-xcode=<a>                    Generate different code for forming addresses; <a>={pic13|pic32}
-xcsi                         Allow C source code using non-ISO C compliant locales
-xdebugformat=<a>             Selects the format of debugging information; <a>={stabs|dwarf}
-xdebuginfo=<a>[,<a>]         Control amount of debugging and observability emitted.;
                              <a>={line,param,variable,tagtype,codetag,macro,decl,hwcprof}
-xdepend[={yes|no}]           Analyze loops for data dependencies
-xdumpmacros[=<a>[,<a>]]      Prints macro definitions on the standard error output;
                              <a>={defs,undefs,use,loc,conds,sys}
-xdryrun                      The same as -###
-xe                           Perform only syntax/semantic checking, no code generation
-xglobalize[={yes|no}]        Control globalization of file static variables and functions.
-xhelp=flags                  Display on-line help information
-xhwcprof[=<a>]               Enable/disable program annotation for hardware counter profiling; <a>={enable|disable}
-xinline=<v>                  Attempt inlining of specified user routines; <v>={%auto,<func>,no%<func>}
-xinline_param=<a>[,<a>]      Fine control of inlining
-xinline_report[=<a>]         Information for tuning inlining; <a>={0|1|2}
-xinstrument=[no_]datarace    Enable/disable instrumentation for race detection tool
-xipo[=<n>]                   Enable optimization and inlining across source files; <n>={0|1|2}
-xipo_archive=<a>             Enable crossfile optimization including archive files;
                              <a>={none|readonly|writeback}
-xipo_build[={yes|no}]        Enable faster build for optimization and inlining cross source files
-xivdep[=<a>]                 Ignore loop-carried dependences on array references in a loop; <a>={loop|loop_any|back|back_any|none}
-xjobs=<n>                    Maximum number of components compiler will fork in parallel
-xkeepframe[=<v>]             Do not optimize stack frame of specified user routine; <v>={%all|%none|[no%]<func>}
-xkeep_unref[=<a>[,<a>]]      Keep definitions of unreferenced functions and variables.;
                              <a>={funcs,vars}
-xlang=<a>                    Control the runtime library behavior of libc; <a>={c99|c11|c89}
-xldscope=<a>                 Indicates the appropriate linker scoping within the source program;
                              <a>={global|symbolic|hidden}
-xlibmieee                    Force IEEE 754 return values for math routines in exceptional cases
-xlibmil                      Inline selected libm math routines for optimization
-xlibmopt                     Link with optimized math library
-xlic_lib=sunperf             (Obsolete) Use -library=sunperf instead
-xlinkopt[={0,1,2}]           Perform optimizations on relocatable object files
-xloopinfo                    Show loops that parallelized
-xmaxopt=[off,1,2,3,4,5]      Maximum optimization level allowed on #pragma opt
-xmodel=<a>                   Specify memory model for 64-bit programs;
                              <a>={small|kernel|medium}
-xnolib                       Do not link with default system libraries
-xnolibmil                    Cancel -xlibmil on command line
-xnolibmopt                   Cancel -xlibmopt on command line
-xnorunpath                   Do not build a runtime search path into the executable
-xopenmp[=<a>]                Enable OpenMP language extension; <a>={none|noopt|parallel}
-xpagesize=<a>                Controls the preferred page size for the stack and for the heap; <a>={4K|2M|4M|1G|default}
-xpagesize_heap=<a>           Controls the preferred page size for the heap; <a>={4K|2M|4M|1G|default}
-xpagesize_stack=<a>          Controls the preferred page size for the stack; <a>={4K|2M|4M|1G|default}
-xpatchpadding[=<a>]          Put space before start of code for hot patching. <a>={fix|patch|<integer>}
-xpch=<t>                     Enable precompiled headers. Collect data for, or use existing, PCH file; <t>={auto|autofirst|{collect,use}:<file>[.cpch]}
-xpchstop=<file>              Specified include file marks end of initial common sequence of pre-processing directives for precompiled headers
-xpec[={yes|no}]              Generate a PEC binary
-xpentium                     Generate code for the pentium processor
-xpg                          Compile for profiling with gprof
-xprefetch[=<p>]              Specify instruction prefetch; <p>={auto,no%auto,explicit,no%explicit}
-xprefetch_auto_type=<a>      Specify automatic indirect prefetch insertion for loops;
                              <a>={indirect_array_access}
-xprefetch_level[=<n>]        Controls the aggressiveness of the -xprefetch=auto option; <n>={1|2|3}
-xprevise[={yes|no}]          Generate static analysis information for the code analyzer
-xprofile=<t>                 Collect data for a profile or use a profile to optimize; <t>={{collect,use}[:<path>],tcov}
-xprofile_ircache[=<t>]       Path to intermediate file cache used with -xprofile option
-xreduction                   Recognize reduction operations in parallelized loops
-xregs=<a>[,<a>]              Specify the usage of optional registers; <a>={float,frameptr}
-xrestrict[=<f>]              Treat pointer valued function parameters as restricted; <f>={%none,%all,<function-name list>}
-xs[={yes|no}]                Allow debugging without object (.o) files
-xsegment_align=<a>           Controls alignment of heap and stack segments;
                              <a>={none|4K|8K|64K|512K|2M|4M|32M|256M|1G}
-xsfpconst                    Represent unsuffixed floating point constants as single precision
-xspace                       Do not do optimizations that increase code size
-xstrconst                    Place string literals into read-only data segment
-xtarget=<a>                  Specify target system for optimization
-xtemp=<dir>                  Set directory for temporary files to <dir>
-xthreadvar[=<a>]             Control code generation for thread variables; <a>={dynamic}
-xtime[=<a>]                  Report the execution time for each compilation phase; <a>={1|2|3}
-xtransition                  Emit warnings for differences between K&R C and ANSI C
-xtrigraphs[={yes|no}]        Enable|disable trigraph translation
-xunboundsym[={yes|no}]       Program contains dynamiclly bound symbols
-xunroll=<n>                  Enable unrolling loops <n> times where possible
-xustr=<a>                    Recognize sixteen-bit string literals; <a>={ascii_utf16_ushort|no}
-xvector[=<a>[,<a>]]          Automatic generation of calls to the vector library functions and/or the generation of the SIMD instructions
-xvpara                       Verbose parallelization warnings

Suffix 'a'              Object library
Suffix 'il'             Inline template file
Suffix 'o'              Object file
Suffix 'so'             Shared object
Suffix 's'              Assembler source
Suffix 'S'              Assembler source for cpp
Suffix 'c'              C source
Suffix 'i'              C source after preprocessing
Suffix 'ln'             Lint library file
Suffix 'd'              Build dependencies file

**Comment by sodabrew** [comment]
I don't see any text in there indicating the name or version of the compiler, and there's no text to match the `/Sun/` regex, which is why that block above ends up thinking you have GCC.

According to http://nadeausoftware.com/articles/2012/10/c_c_tip_how_detect_compiler_name_and_version_using_compiler_predefined_macros perhaps we could do a compiler test to check for the `__SUNPRO_CC` macro.

On the flip side, we could check if the compiler is gcc by checking `#{RbConfig::CONFIG['CC']} --version` and assume it is a Sun compiler if that doesn't work.

**Comment by CirrusThink** [comment]
They both sound like good options, we still have this issue though when compiling with CC & cc with -Kpic

Unable to load the EventMachine C extension; To use the pure-ruby reactor, require 'em/pure_ruby'
/export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/gems/2.1.0/bundler/gems/eventmachine-2333335a9fbf/lib/eventmachine.rb:8:in `require': ld.so.1: ruby: fatal: relocation error: R_AMD64_PC32: file /export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/gems/2.1.0/bundler/gems/extensions/x86_64-solaris-2.11/2.1.0-static/eventmachine-2333335a9fbf/rubyeventmachine.so: symbol _init: value 0x7f0044410225 does not fit - /export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/gems/2.1.0/bundler/gems/extensions/x86_64-solaris-2.11/2.1.0-static/eventmachine-2333335a9fbf/rubyeventmachine.so (LoadError)
        from /export/home/dolly/ruby64/ruby-2.1.6/lib/ruby/gems/2.1.0/bundler/gems/eventmachine-2333335a9fbf/lib/eventmachine.rb:8:in `<top (required)>'
        from ./initializer.rb:8:in `require'
        from ./initializer.rb:8:in `<main>'

**Comment by sodabrew** [comment]
Does it work if you compile without KPIC?

**Comment by CirrusThink** [comment]
I believe its the same issue. I will have time to try again later this week end end post an update

**Comment by sodabrew** [comment]
I'd like to get a fix for this into a future release. Have you been able to narrow down either the detection of Studio compiler, or the issue with KPIC?

**Comment by sodabrew** [comment]
... so I just downloaded a Solaris VM and got Studio 12.4 installed. I hit paydirt:
```
$ /opt/solarisstudio12.4/bin/cc -V
cc: Sun C 5.13 SunOS i386 2014/10/20
```

I'll update the `extconf.rb` to use `cc -V` to discover the Studio compiler. I'm still not sure what to do about the mismatched CC / CXX variables. It looks like Studio doesn't have a separate C++ compiler binary?

**Comment by CirrusThink** [comment]
Uhm the only other thing here is that although you get a -V i386 you are most likely using a x64 installation of Solaris.

**Comment by sodabrew** [comment]
Yes, I have a 32-bit Ruby on a 64-bit Solaris. I didn't expect that default, but now I recall Sun was always big on 32-but user spaces. Looking forward to your feedback!

**Comment by CirrusThink** [comment]
"Yes, I have a 32-bit Ruby on a 64-bit Solaris." is this via rvm ? the default ruby on Solaris 11 11.1 and 11.2 is 64 bit. RVM is also upgrading their software to recognize this issue so that a 64 bit version can be correctly installed. I believe they added it to their master a couple of days ago.

**Comment by sodabrew** [comment]
I use rbenv and build-ruby, same thing but different.

**Comment by CirrusThink** [comment]
Same issue. They are building your ruby off of uname

Sent from my iPhone

Anthony Palermo
C.E.O. & Co-Founder 

Twitter: @dollydrive
Facebook DollyDrive

**Comment by sodabrew** [comment]
Ok, but, bottom line is current master working or not working for you?

**Comment by CirrusThink** [comment]
Donno yet I will set time this week to test it out and update ticket 



Anthony Palermo 
Co-Founder 

Phone: +1 (305) 303-3204 
Twitter: @dollydrive 
Facebook DollyDrive 
Skype aderium 

----- Original Message -----
